feat: init esnekpos connector

adnanjpg opened the pull request:
## Type of Change
<!-- Put an `x` in the boxes that apply -->

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
This PR introduces a new payment connector [`esnekpos`](https://developer-eng.esnekpos.com/), which is a Turkish payment provider.

At this stage, only the payment authorization api service is functioning, I'm expecting feedback on the implementation so I would continue adding the other apis for this connector and implementing a few other Turkish connectors.


### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

<!--
Provide links to the files with corresponding changes.

Following are the paths where you can find config files:
1. `config`
2. `crates/router/src/configs`
3. `loadtest/config`
-->


## Motivation and Context
<!--
Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.

If you don't have an issue, we'd recommend starting with one first so the PR
can focus on the implementation (unless it is an obvious bug or documentation fix
that will have little conversation).
-->


## How did you test it?
<!--
Did you write an integration/unit/API test to verify the code changes?
Or did you test this change manually (provide relevant screenshots)?
-->


## Checklist
<!-- Put an `x` in the boxes that apply -->

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [x] I added unit tests for my changes where possible

adnanjpg commented:
There's also an implementation for Esnekpos for the hyperswitch-control-center project, that will be submitted once this PR is ready

SanchithHegde commented:
Hey @adnanjpg, can you please base your PR on the `main` branch instead if the `v1.108.2` tag?

adnanjpg commented:
Hi @SanchithHegde , I've based my PR on the main branch, but now my `esnekpos.rs` file has the following error:

```
the trait bound `connector::esnekpos::Esnekpos: hyperswitch_interfaces::api::ConnectorIntegration<hyperswitch_domain_models::router_flow_types::PostProcessing, PaymentsPostProcessingData, PaymentsResponseData>` is not satisfied
```

and I'm unable to figure out why

Sakilmostak commented:
Hello @adnanjpg , you need to go to `crates/router/src/core/payments/flows.rs` and put your connector name in the `default_imp_for_pre_processing_steps` macro. This should work out👍

adnanjpg commented:
Hi @Sakilmostak , the macro mentioned already has my connector name

a0m0rajab commented:
Related issue: https://github.com/juspay/hyperswitch/issues/5833

Sakilmostak commented on crates/router/src/core/payments/flows.rs:
```diff
@@ -1022,6 +1031,7 @@ impl<const T: u8>
 
 default_imp_for_pre_processing_steps!(
     connector::Adyenplatform,
+    connector::Esnekpos,
```
It seems you have filled for `default_imp_for_pre_processing_steps` whereas you need to fill for `default_imp_for_post_processing_steps`

adnanjpg commented:
Hi @Sakilmostak , I just fixed that, please check again

Sakilmostak commented:
On further inspecting your branch, I found that it is the same case with Authorize Session Token flow. Please add the connector name in the `default_imp_for_authorize_session_token`. 
And it is also missing in whole v2 flow suite. Please add in the respective implementations in `connector_integration_v2_impls.rs` file

adnanjpg commented:
Hi @Sakilmostak , I think these commits resolve your requests

I've also fixed merge conflicts with main

Sakilmostak commented on crates/router/src/connector/esnekpos/transformers.rs:
```diff
@@ -0,0 +1,531 @@
+    product: Vec<EsnekposPaymentRequestProduct>,
+}
+
+#[derive(Default, Debug, Serialize, Eq, PartialEq)]
+pub struct EsnekposCard {
+    #[serde(rename = "CC_NUMBER")]
+    number: cards::CardNumber,
+    #[serde(rename = "EXP_MONTH")]
+    expiry_month: Secret<String>,
+    #[serde(rename = "EXP_YEAR")]
+    expiry_year: Secret<String>,
+    #[serde(rename = "CC_CVV")]
+    cvc: Secret<String>,
+    #[serde(rename = "CC_OWNER")]
+    holder_name: Secret<String>,
+    #[serde(rename = "INSTALLMENT_NUMBER")]
+    installment_number: Secret<i32>,
+    complete: bool,
+}
```
Please do the same here

Sakilmostak commented on crates/router/src/connector/esnekpos/transformers.rs:
```diff
@@ -0,0 +1,531 @@
+    }
+}
+
+#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq)]
+pub enum EsnekposPaymentStatus {
+    #[serde(rename = "SUCCESS")]
+    Succeeded,
+    #[serde(rename = "ERROR")]
+    Error,
+    #[default]
+    #[serde(rename = "FIELD_ERROR")]
+    FieldError,
+    #[serde(rename = "AUTHENTICATION_ERROR")]
+    AuthenticationError,
+    #[serde(rename = "LIMIT_ERROR")]
+    LimitError,
+    #[serde(rename = "COMMISSION_ERROR")]
+    CommissionError,
+    #[serde(rename = "INSERT_ERROR")]
+    InsertError,
+    #[serde(rename = "BRAND_ERROR")]
+    BrandError,
+    #[serde(rename = "PAYMENT_ERROR")]
+    PaymentError,
+    #[serde(rename = "NOT_3D_AUTHENTICATION")]
+    Not3DAuthentication,
+    #[serde(rename = "BLOCKED_ERROR")]
+    BlockedError,
+    #[serde(rename = "ORDER_CANCEL")]
+    OrderCancel,
+    #[serde(rename = "PROCESS_QUERY")]
+    ProcessQuery,
+    #[serde(rename = "DIRECT_PAYMENT")]
+    DirectPayment,
+}
```
Please do the same here

Sakilmostak commented on crates/router/src/connector/esnekpos/transformers.rs:
```diff
@@ -0,0 +1,531 @@
+    pub return_code: String,
+    #[serde(rename = "RETURN_MESSAGE")]
+    pub return_message: String,
+    #[serde(rename = "RETURN_MESSAGE_TR")]
+    pub return_message_tr: Option<String>,
+    #[serde(rename = "ERROR_CODE")]
+    pub error_code: Option<String>,
+    #[serde(rename = "AUTH_HASH")]
+    pub auth_hash: Option<String>,
+    #[serde(rename = "BANK_AUTH_CODE")]
+    pub bank_auth_code: Option<String>,
+    #[serde(rename = "DATE")]
+    pub date: String,
+    #[serde(rename = "URL_3DS")]
+    pub url_3ds: Option<String>,
+    #[serde(rename = "REFNO")]
+    pub refno: String,
+    #[serde(rename = "HASH")]
+    pub hash: String,
+    #[serde(rename = "COMMISSION_RATE")]
+    pub commission_rate: String,
+    #[serde(rename = "CUSTOMER_NAME")]
+    pub customer_name: String,
+    #[serde(rename = "CUSTOMER_MAIL")]
+    pub customer_mail: String,
+    #[serde(rename = "CUSTOMER_PHONE")]
+    pub customer_phone: String,
+    #[serde(rename = "CUSTOMER_ADDRESS")]
+    pub customer_address: Option<String>,
+    #[serde(rename = "CUSTOMER_CC_NUMBER")]
+    pub customer_cc_number: String,
+    #[serde(rename = "CUSTOMER_CC_NAME")]
+    pub customer_cc_name: Option<String>,
+    #[serde(rename = "IS_NOT_3D_PAYMENT")]
+    pub is_not_3d_payment: bool,
+    #[serde(rename = "VIRTUAL_POS_VALUES")]
+    pub virtual_pos_values: Option<String>,
+    #[serde(rename = "RETURN_MESSAGE_3D")]
+    pub return_message_3d: Option<String>,
+}
```
Please do the same here

Sakilmostak commented on crates/router/src/connector/esnekpos/transformers.rs:
```diff
@@ -0,0 +1,531 @@
+            RefundStatus::Succeeded => Self::Success,
+            RefundStatus::Failed => Self::Failure,
+            RefundStatus::Processing => Self::Pending,
+            //TODO: Review mapping
```
Remove todo!

Sakilmostak commented on crates/router/src/connector/esnekpos/transformers.rs:
```diff
@@ -0,0 +1,531 @@
+    }
+}
+
+//TODO: Fill the struct with respective fields
```
Remove todo!

Sakilmostak commented on crates/router/src/connector/esnekpos/transformers.rs:
```diff
@@ -0,0 +1,531 @@
+    }
+}
+
+#[derive(Default, Debug, Serialize, Deserialize, PartialEq)]
+pub struct EsnekposErrorResponse {
+    #[serde(rename = "RETURN_CODE")]
+    pub code: String,
+    #[serde(rename = "RETURN_MESSAGE")]
+    pub message: String,
+    #[serde(rename = "STATUS")]
+    pub status: Option<String>,
+}
```
Please do the same here

Sakilmostak commented on crates/router/src/core/payments/flows.rs:
```diff
@@ -2023,6 +2044,7 @@ impl<const T: u8>
 
 default_imp_for_reject!(
     connector::Adyenplatform,
+    connector::Esnekpos,
```
Please arrange this alphabetically

Sakilmostak commented on crates/router/src/core/payments/flows.rs:
```diff
@@ -2096,6 +2118,7 @@ impl<const T: u8> api::FraudCheck for connector::DummyConnector<T> {}
 
 default_imp_for_fraud_check!(
     connector::Adyenplatform,
+    connector::Esnekpos,
```
Please arrange this alphabetically

Sakilmostak commented on crates/router/src/core/payments/flows.rs:
```diff
@@ -2202,6 +2225,7 @@ impl<const T: u8>
 #[cfg(feature = "frm")]
 default_imp_for_frm_sale!(
     connector::Adyenplatform,
+    connector::Esnekpos,
```
Please arrange this alphabetically

Sakilmostak commented on crates/router/tests/connectors/sample_auth.toml:
```diff
@@ -258,6 +258,30 @@ api_secret = "Secret key"
+[esnekpos]
+api_key="Merchant Key"
+key1="Merchant"
+
+[fiservemea]
+api_key="API Key"
+
+[nexixpay]
+api_key="API Key"
+
+[wellsfargopayout]
+api_key = "Consumer Key"
+key1 = "Gateway Entity Id"
+api_secret = "Consumer Secret"
+
+[taxjar]
+api_key = "API Key"
+
+[novalnet]
+api_key="API Key"
+
+[deutschebank]
+api_key="API Key"
+
```
remove these clone

Sakilmostak commented on docker-compose-development.yml:
```diff
@@ -337,4 +337,4 @@ services:
       envSdkUrl: http://localhost:9050
       envBackendUrl: http://localhost:8080
-      envLoggingUrl: http://localhost:8207
+      envLoggingUrl: http://localhost:8207
```
revert this change

adnanjpg replied:
I tried to revert this, I thought it was about line endings and tried a couple of things but didn't work

Sakilmostak commented on docker-compose.yml:
```diff
@@ -412,4 +412,4 @@ services:
     volumes:
       - ./config/vector.yaml:/etc/vector/vector.yaml
-      - /var/run/docker.sock:/var/run/docker.sock
+      - /var/run/docker.sock:/var/run/docker.sock
```
revert this change

Sakilmostak commented on crates/router/src/connector/esnekpos.rs:
```diff
@@ -0,0 +1,386 @@
+        _req: &types::PaymentsSyncRouterData,
+        _connectors: &settings::Connectors,
+    ) -> CustomResult<Option<services::Request>, errors::ConnectorError> {
+        Err(errors::ConnectorError::NotImplemented(
+            "ConnectorIntegration<api::PSync build_request method".to_string(),
+        )
+        .into())
```
Can you please throw `FlowNotSupported` instead with flow as `Payment Sync` and connector as `esnekpos`

Sakilmostak commented on crates/router/src/connector/esnekpos.rs:
```diff
@@ -0,0 +1,579 @@
+                .method(services::Method::Post)
+                .url(&types::PaymentsCaptureType::get_url(self, req, connectors)?)
+                .attach_default_headers()
+                .headers(types::PaymentsCaptureType::get_headers(
+                    self, req, connectors,
+                )?)
+                .set_body(types::PaymentsCaptureType::get_request_body(
+                    self, req, connectors,
+                )?)
+                .build(),
+        ))
+    }
+
+    fn handle_response(
+        &self,
+        data: &types::PaymentsCaptureRouterData,
+        event_builder: Option<&mut ConnectorEvent>,
+        res: Response,
+    ) -> CustomResult<types::PaymentsCaptureRouterData, errors::ConnectorError> {
+        let response: esnekpos::EsnekposPaymentsResponse = res
+            .response
+            .parse_struct("Esnekpos PaymentsCaptureResponse")
+            .change_context(errors::ConnectorError::ResponseDeserializationFailed)?;
+        event_builder.map(|i| i.set_response_body(&response));
+        router_env::logger::info!(connector_response=?response);
+        types::RouterData::try_from(types::ResponseRouterData {
+            response,
+            data: data.clone(),
+            http_code: res.status_code,
+        })
+    }
+
+    fn get_error_response(
+        &self,
+        res: Response,
+        event_builder: Option<&mut ConnectorEvent>,
+    ) -> CustomResult<ErrorResponse, errors::ConnectorError> {
+        self.build_error_response(res, event_builder)
+    }
+}
```
Please do the given changes

Sakilmostak commented on crates/router/src/connector/esnekpos.rs:
```diff
@@ -0,0 +1,386 @@
+        _req: &types::RefundsRouterData<api::Execute>,
+        _connectors: &settings::Connectors,
+    ) -> CustomResult<Option<services::Request>, errors::ConnectorError> {
+        Err(errors::ConnectorError::NotImplemented(
+            "ConnectorIntegration<api::Execute build_request method".to_string(),
+        )
+        .into())
+    }
```
Please throw the `FlowNotSupported` error with flow as `Refunds` and connector `esnekpos`

Sakilmostak commented on crates/router/src/connector/esnekpos.rs:
```diff
@@ -0,0 +1,386 @@
+        _req: &types::RefundSyncRouterData,
+        _connectors: &settings::Connectors,
+    ) -> CustomResult<Option<services::Request>, errors::ConnectorError> {
+        Err(errors::ConnectorError::NotImplemented(
+            "ConnectorIntegration<api::RSync build_request method".to_string(),
+        )
+        .into())
```
Please throw the `FlowNotSupported` error with flow as `Refund Sync` and connector `esnekpos`

Sakilmostak commented on crates/router/src/connector/esnekpos/transformers.rs:
```diff
@@ -0,0 +1,385 @@
+                                });
+                                let customer = match customer_res {
+                                    Ok(customer) => customer,
+                                    Err(e) => return Err(e.into()),
+                                };
+
+                                let auth = EsnekposAuthType::try_from(
+                                    &item.router_data.connector_auth_type,
+                                )?;
+
+                                Ok(Self {
+                                    config: EsnekposPaymentRequestConfig {
+                                        merchant: auth.merchant.expose(),
+                                        merchant_key: auth.merchant_key.expose(),
+                                        back_url: item.router_data.request.get_return_url()?,
+                                        order_ref_number: item.router_data.payment_id.clone(),
+                                        prices_currency: item
+                                            .router_data
+                                            .request
+                                            .currency
+                                            .to_string(),
+                                        order_amount: item.amount.get_amount_as_string(),
+                                    },
+                                    credit_card,
+                                    customer,
+                                    product: vec![],
+                                })
+                            }
+                        }
+                    }
+                    _ => Err(
+                        errors::ConnectorError::NotImplemented("Payment methods".to_string())
+                            .into(),
+                    ),
+                }
+            }
+            Err(e) => Err(e.into()),
+        }
+    }
+}
```
Please don't nest the match statements, it reduces the readability of code, rather try to extract the data that are needed separately and and then construct the struct. For help, you can use utils function present in `crates/router/src/connector/utils.rs` for such extraction of data. Please go through any connector implementation as an example to construct in the following way. For example, you can go through try_from implementation for `AdyenPaymentRequest` in `crates/router/src/connector/adyen/transformers.rs` for an understanding.

Sakilmostak reviewed (changes requested):
Otherwise LGTM!

hrithikesh026 reviewed (dismissed):
Core changes Looks good to me

Gnanasundari24 commented:
@adnanjpg Please resolve the conflicts and add the testcases.

adnanjpg commented:
Hi @Gnanasundari24 , I've just fixed the conflicts
about the tests, I have added the default tests by copying from `ebanx`, if more specific tests are needed can you please give me more details?

adnanjpg commented:
Hi, is there a way we can fast track this? I'm having to continually deal with merge conflicts as my pr affects core parts of the codebase

AkshayaFoiger commented on crates/router/src/connector/esnekpos/transformers.rs:
```diff
@@ -0,0 +1,263 @@
+                    cc_cvc: req_card.card_cvc,
+                    cc_owner: card_holder_name,
+                    // TODO(adnanjpg): get the installment number from the request
+                    installment_number: Secret::new(0),
```
Instead of using a magic number, define a variable and  as Hyperswitch does not support installments, it's okay to set its value to 0, with relevant comment.  
This is not a sensitive field

AkshayaFoiger commented on crates/router/src/connector/esnekpos/transformers.rs:
```diff
@@ -0,0 +1,263 @@
+                    },
+                    credit_card,
+                    customer,
+                    product: vec![],
```
Is this a mandatory field ?

AkshayaFoiger commented on crates/router/src/connector/esnekpos/transformers.rs:
```diff
@@ -0,0 +1,263 @@
+        item: types::ResponseRouterData<F, EsnekposPaymentResponse, T, types::PaymentsResponseData>,
+    ) -> Result<Self, Self::Error> {
+        Ok(Self {
+            status: enums::AttemptStatus::from(item.response.status),
```
Incase of failure status, return a `ErrorResponse`

AkshayaFoiger commented on crates/router/src/connector/esnekpos/transformers.rs:
```diff
@@ -0,0 +1,263 @@
+                connector_metadata: None,
+                network_txn_id: None,
+                connector_response_reference_id: None,
+                incremental_authorization_allowed: None,
```
If `order_ref_number` is returned in transaction response, map it to `connector_response_reference_id`

AkshayaFoiger commented on crates/router/src/connector/esnekpos/transformers.rs:
```diff
@@ -0,0 +1,263 @@
+    pub return_message_3d: Option<String>,
+}
+
+impl<F, T>
+    TryFrom<types::ResponseRouterData<F, EsnekposPaymentResponse, T, types::PaymentsResponseData>>
+    for types::RouterData<F, T, types::PaymentsResponseData>
+{
+    type Error = error_stack::Report<errors::ConnectorError>;
+    fn try_from(
+        item: types::ResponseRouterData<F, EsnekposPaymentResponse, T, types::PaymentsResponseData>,
+    ) -> Result<Self, Self::Error> {
+        Ok(Self {
+            status: enums::AttemptStatus::from(item.response.status),
+            response: Ok(types::PaymentsResponseData::TransactionResponse {
+                resource_id: types::ResponseId::ConnectorTransactionId(item.response.refno),
+                redirection_data: Box::new(None),
+                mandate_reference: Box::new(None),
+                connector_metadata: None,
+                network_txn_id: None,
+                connector_response_reference_id: None,
+                incremental_authorization_allowed: None,
+                charge_id: None,
+            }),
+            ..item.data
+        })
+    }
+}
+
```
Could you explain the complete payment flow? It appears from the [connector doc](https://developer-eng.esnekpos.com/esnekpos-developer-environment/payment-service/3d-payment#here-is-a-sample-json-file-to-send-to-the-service) that you need to construct `redirection_data`, but this is not implemented in the code.
